Penteract/Assets/Scripts/AIMeleeGrunt.cpp

```diff
@@ -251,152 +257,175 @@ void AIMeleeGrunt::OnAnimationEvent(StateMachineEnum stateMachineEnum, const cha
-                    animation->SendTrigger("RunDeath1");
-                }
-                else {
-                    animation->SendTrigger("RunDeath2");
-                }
-            }
-			else if (state == AIState::STUNNED) {
-				if (deathType) {
-					animation->SendTrigger("StunnedDeath1");
-				}
-				else {
-					animation->SendTrigger("StunnedDeath2");
-				}
-			}
-			else if (state == AIState::PUSHED) {
-				if (deathType) {
-					animation->SendTrigger("HurtDeath1");
-				}
-				else {
-					animation->SendTrigger("HurtDeath2");
-				}
-			}
-
-			if (audios[static_cast<int>(AudioType::DEATH)]) audios[static_cast<int>(AudioType::DEATH)]->Play();
-			ComponentCapsuleCollider* collider = GetOwner().GetComponent<ComponentCapsuleCollider>();
-			if (collider) collider->Disable();
-
-			agent->RemoveAgentFromCrowd();
-			if (gruntCharacter.beingPushed) gruntCharacter.beingPushed = false;
-			state = AIState::DEATH;
-
-		}
-	}
-}
-
-void AIMeleeGrunt::EnableBlastPushBack() {
-	if (state != AIState::START && state != AIState::SPAWN && state != AIState::DEATH) {
-		gruntCharacter.beingPushed = true;
-		state = AIState::PUSHED;
-		if (animation->GetCurrentState()) animation->SendTrigger(animation->GetCurrentState()->name + "Hurt");
-	}
-}
-
-void AIMeleeGrunt::DisableBlastPushBack() {
-	if (state != AIState::START && state != AIState::SPAWN && state != AIState::DEATH) {
-		gruntCharacter.beingPushed = false;
-		if (animation->GetCurrentState()) animation->SendTrigger(animation->GetCurrentState()->name + "Idle");
-		state = AIState::IDLE;
-	}
-}
-
-bool AIMeleeGrunt::IsBeingPushed() const {
-	return gruntCharacter.beingPushed;
-}
-
-void AIMeleeGrunt::UpdatePushBackPosition() {
-	float3 playerPos = player->GetComponent<ComponentTransform>()->GetGlobalPosition();
-	float3 enemyPos = GetOwner().GetComponent<ComponentTransform>()->GetGlobalPosition();
-	float3 initialPos = enemyPos;
-
-	float3 direction = (enemyPos - playerPos).Normalized();
-
-	if (agent) {
-		enemyPos += direction * gruntCharacter.pushBackSpeed * Time::GetDeltaTime();
-		agent->SetMoveTarget(enemyPos, false);
-		float distance = enemyPos.Distance(initialPos);
-		currentPushBackDistance += distance;
-
-		if (currentPushBackDistance >= gruntCharacter.pushBackDistance) {
-			DisableBlastPushBack();
-			currentPushBackDistance = 0.f;
-		}
-	}
-}
+}
+
+void AIMeleeGrunt::OnCollision(GameObject& collidedWith, float3 collisionNormal, float3 penetrationDistance, void* particle) {
+	if (state != AIState::START && state != AIState::SPAWN) {
+		if (gruntCharacter.isAlive && playerController) {
+			bool hitTaken = false;
+			if (collidedWith.name == "FangBullet") {
+				if (!particle) return;
+				ComponentParticleSystem::Particle* p = (ComponentParticleSystem::Particle*)particle;
+				ComponentParticleSystem* pSystem = collidedWith.GetComponent<ComponentParticleSystem>();
+				GameplaySystems::DestroyGameObject(&collidedWith);
+				if (pSystem) pSystem->KillParticle(p);

```

Comment:
Wouldn't it be better to first delete the particle and then call DestroyGameObject?